### ISSUE TITLE
TwistedProtocolConnection references connectTCP/connectSSL, but should reference endpoints

### ISSUE THREAD

**Issue by glyph** [pika/pika#658]
connectTCP/connectSSL are low level methods which should only be of interest to reactor implementors.  Instead, refer people to endpoints: https://twistedmatrix.com/documents/15.4.0/core/howto/endpoints.html

**Comment by vitaly-krugl** [comment]
Could someone with Twisted expertise tackle this one?

**Comment by jeremycline** [comment]
I have been interested in using Pika with Twisted's endpoints API, but I've run into trouble. The script I've tried is:
```
from pika.adapters import twisted_connection
from twisted.internet import endpoints, protocol, reactor
from twisted.application import service
from twisted.application.internet import ClientService

amqp_endpoint = endpoints.clientFromString(reactor, 'tcp:host=127.0.0.1:port=5672:timeout=15')
amqp_factory = protocol.Factory.forProtocol(
    twisted_connection.TwistedProtocolConnection)
amqp_service = ClientService(amqp_endpoint, amqp_factory)

application = service.Application('Twisted Pika Client')
service_collection = service.IServiceCollection(application)
amqp_service.setServiceParent(service_collection)
```
The problem is, Twisted keeps reconnecting:
```
 jcline  ⓔ  twisted  ~  devel  playpen  twistd -n -y pika_twisted.py 
2016-10-30T10:54:37-0400 [-] Loading pika_twisted.py...
2016-10-30T10:54:37-0400 [-] Loaded.
2016-10-30T10:54:37-0400 [twisted.scripts._twistd_unix.UnixAppLogger#info] twistd 16.4.1 (/home/jcline/.virtualenvs/twisted/bin/python 2.7.12) starting up.
2016-10-30T10:54:37-0400 [twisted.scripts._twistd_unix.UnixAppLogger#info] reactor class: twisted.internet.epollreactor.EPollReactor.
2016-10-30T10:54:37-0400 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7fce27cbf710>
2016-10-30T10:54:37-0400 [twisted.application.internet.ClientService#info] Scheduling retry 1 to connect <twisted.internet.endpoints.TCP4ClientEndpoint object at 0x7fce27d06150> in 1.69262447035 seconds.
2016-10-30T10:54:37-0400 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7fce27cbf710>
2016-10-30T10:54:39-0400 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7fce27cbf710>
2016-10-30T10:54:39-0400 [twisted.application.internet.ClientService#info] Scheduling retry 2 to connect <twisted.internet.endpoints.TCP4ClientEndpoint object at 0x7fce27d06150> in 2.96890088608 seconds.
2016-10-30T10:54:39-0400 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7fce27cbf710>
^C2016-10-30T10:54:41-0400 [-] Received SIGINT, shutting down.
2016-10-30T10:54:41-0400 [-] Main loop terminated.
2016-10-30T10:54:41-0400 [twisted.scripts._twistd_unix.UnixAppLogger#info] Server Shut Down.
```

When I tcpdump the conversation, I see something rather unexpected. The handshake goes fine, instantly after Twisted finishes the handshake it sends a FIN and the connection is reset. The logs aren't very informative about what went wrong, so I looked at the process with a debugger and Twisted is raising a Timeout error, even though the timeout was not hit at all.

This is with pika-0.10.0 and twisted-16.4.1. I'm at a bit of a loss since I don't know the Twisted internals that well, so I can't say whether this is a problem with Pika or with Twisted. @glyph does this ring any bells?

**Comment by glyph** [comment]
@jeremycline Can you create a minimal reproducer of this example?  One without using Pika, for example? :).  I don't know much about how `twisted_connection` works, but this shouldn't be happening; at the very least it should be logging more information about why it was disconnected.

**Comment by glyph** [comment]
@jeremycline If you use a different timeout value, does it work? Alternately; if you eliminate the timeout entirely, does that work? And finally, how about `HostnameEndpoint` with localhost?

**Comment by jeremycline** [comment]
@glyph Using different timeout values (I tried 30 and 60) or removing the timeout entirely does not change the behaviour.

Switching to ``amqp_endpoint = endpoints.HostnameEndpoint(reactor, 'localhost', 5672)`` changed the log output to
```
2016-10-31T14:12:01+0000 [-] Loading pika_twisted.py...
2016-10-31T14:12:01+0000 [-] Loaded.
2016-10-31T14:12:01+0000 [twisted.scripts._twistd_unix.UnixAppLogger#info] twistd 16.5.0 (/home/vagrant/.virtualenvs/pika/bin/python2 2.7.12) starting up.
2016-10-31T14:12:01+0000 [twisted.scripts._twistd_unix.UnixAppLogger#info] reactor class: twisted.internet.epollreactor.EPollReactor.
2016-10-31T14:12:01+0000 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:01+0000 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:02+0000 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:02+0000 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:02+0000 [twisted.application.internet.ClientService#info] Scheduling retry 1 to connect <twisted.internet.endpoints.HostnameEndpoint object at 0x7f4c62110450> in 2.12706868345 seconds.
2016-10-31T14:12:04+0000 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:04+0000 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:04+0000 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:04+0000 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:05+0000 [twisted.application.internet.ClientService#info] Scheduling retry 2 to connect <twisted.internet.endpoints.HostnameEndpoint object at 0x7f4c62110450> in 2.55572051894 seconds.
2016-10-31T14:12:07+0000 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:07+0000 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:07+0000 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:07+0000 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7f4c620baea8>
2016-10-31T14:12:08+0000 [twisted.application.internet.ClientService#info] Scheduling retry 3 to connect <twisted.internet.endpoints.HostnameEndpoint object at 0x7f4c62110450> in 4.11712648038 seconds.
```

It turns out this is because it's trying ::1 and then 127.0.0.1, both of which are doing the handshake followed immediately by FIN.

I set up a duplicate experiment using the mail client example[0] as a base, and that did not exhibit the same behaviour. The only way this differs from the pika example I posted is the protocol class I'm providing to ``forProtocol``:
```
from twisted.internet import endpoints, protocol, reactor
from twisted.application import service
from twisted.application.internet import ClientService
from twisted.mail import imap4


email_endpoint = endpoints.clientFromString(reactor, 'tcp:host=127.0.0.1:port=2525:timeout=15')
email_factory = protocol.Factory.forProtocol(imap4.IMAP4Client)
email_service = ClientService(email_endpoint, email_factory)

application = service.Application('Twisted Pika Client')
service_collection = service.IServiceCollection(application)
email_service.setServiceParent(service_collection)
```

The logs from running this:
```
2016-10-31T14:21:03+0000 [-] Loading email.py...
2016-10-31T14:21:03+0000 [-] Loaded.
2016-10-31T14:21:03+0000 [twisted.scripts._twistd_unix.UnixAppLogger#info] twistd 16.5.0 (/home/vagrant/.virtualenvs/pika/bin/python2 2.7.12) starting up.
2016-10-31T14:21:03+0000 [twisted.scripts._twistd_unix.UnixAppLogger#info] reactor class: twisted.internet.epollreactor.EPollReactor.
2016-10-31T14:21:03+0000 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7f49407847e8>
^C2016-10-31T14:22:18+0000 [-] Received SIGINT, shutting down.
2016-10-31T14:22:18+0000 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7f49407847e8>
2016-10-31T14:22:18+0000 [-] Main loop terminated.
2016-10-31T14:22:18+0000 [twisted.scripts._twistd_unix.UnixAppLogger#info] Server Shut Down.
```

To take Rabbit out of the equation I just started listening on a socket with ``nc -l``. In the email example, the connection in left open and ``nc`` continues to listen, but in the pika example ``nc`` immediately exits when it closes the connection.

I'm speculating wildly here, since I don't know much about the internals of either project (yet), but could it be that Twisted hands the established connection over to Pika to do any initial setup it needs to do and pika is somehow indicating (incorrectly) to Twisted that it couldn't do its setup, so Twisted is cleaning the TCP connection up and starting over?

[0] http://twistedmatrix.com/trac/#imap4client

**Comment by glyph** [comment]
This… might just be an error-reporting bug.

`protocol.Factory.forProtocol`, formally, takes a 0-argument callable that returns an `IProtocol` provider; idiomatically, that's a `Protocol` subclass with a constructor that takes no argument.

What you're passing, `TwistedChannel`, appears to have a totally different interface; for starters, it takes a required argument.  That alone will make it blow up during `buildProtocol`.

This is a bug in Twisted: your logs ought to be clearly showing a traceback!  Please feel free to file this at https://twistedmatrix.com/

However, the fix for you should be simple.  Write a function that takes no arguments and returns a `TwistedProtocolConnection` (which _is_ a subclass of `Protocol`, although its constructor still requires a "`parameters`" argument) where you're passing `TwistedChannel` right now.

Does that make sense?

**Comment by jeremycline** [comment]
@glyph Fantastic, altering ``twisted_connection.TwistedProtocolConnection`` to not require ``parameters`` got this working. I'll file the issue for the error-reporting problem in Twisted. 

Unfortunately, I _still_ can't get this to behave like I'd expect. If I try to get the active client with [ClientService.whenConnected()](https://twistedmatrix.com/documents/current/core/howto/endpoints.html#getting-the-active-client), it works just fine the first time around, but if RabbitMQ goes down, calling this gets me the same ``TwistedProtocolConnection`` instance, which is disconnected:
```

from pika.adapters import twisted_connection
from twisted.internet import defer, endpoints, protocol, reactor
from twisted.application import service
from twisted.application.internet import ClientService

amqp_endpoint = endpoints.clientFromString(
    reactor, 'tcp:host=localhost:port=5672:timeout=15')

amqp_factory = protocol.Factory.forProtocol(
    twisted_connection.TwistedProtocolConnection)

amqp_service = ClientService(amqp_endpoint, amqp_factory)
application = service.Application('Twisted Pika client')
service_collection = service.IServiceCollection(application)
amqp_service.setServiceParent(service_collection)


@defer.inlineCallbacks
def setup(amqp_service):
    """Returns a Deferred that fires with the queue and consumer tag when it's ready"""
    connection = yield amqp_service.whenConnected()
    yield connection.ready
    channel = yield connection.channel()
    yield channel.exchange_declare(exchange='topic_link', type='topic')
    yield channel.queue_declare(
        queue='hello', auto_delete=False, exclusive=False)
    yield channel.queue_bind(
        exchange='topic_link', queue='hello', routing_key='hello.world')
    yield channel.basic_qos(prefetch_count=1)
    queue_object, consumer_tag = yield channel.basic_consume(
        queue='hello', no_ack=False)

    defer.returnValue((queue_object, consumer_tag))


@defer.inlineCallbacks
def read(amqp_service):
    """Consumes messages forever.

    This checks for an active connection each time, so you can stop and start
    the AMQP server and this will continue to work.
    """
    queue_object, consumer_tag = yield setup(amqp_service)
    while True:
        # Get a single message and print it
        try:
            channel, method, properties, body = yield queue_object.get()
            if body:
                print(body)
            yield channel.basic_ack(delivery_tag=method.delivery_tag)
        except Exception as e:
            print(repr(e))
            queue_object, consumer_tag = yield setup(amqp_service)


read(amqp_service)
```

This script results in:

```
2016-11-01T19:54:30+0000 [stdout#info] The channel was closed (320) CONNECTION_FORCED - broker forced connection closure with reason 'shutdown'
2016-11-01T19:54:30+0000 [twisted.internet.defer#critical] Unhandled error in Deferred:
2016-11-01T19:54:30+0000 [twisted.internet.defer#critical] 
        Traceback (most recent call last):
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/twisted/internet/defer.py", line 1258, in _inlineCallbacks
            result = result.throwExceptionIntoGenerator(g)
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
            return g.throw(self.type, self.value, self.tb)
          File "test.py", line 60, in read
            queue_object, consumer_tag = yield setup(amqp_service)
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/twisted/internet/defer.py", line 1406, in unwindGenerator
            return _inlineCallbacks(None, gen, Deferred())
        --- <exception caught here> ---
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/twisted/internet/defer.py", line 1260, in _inlineCallbacks
            result = g.send(result)
          File "test.py", line 29, in setup
            channel = yield connection.channel()
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/pika/adapters/twisted_connection.py", line 416, in channel
            base_connection.BaseConnection.channel(self, d.callback, channel_number)
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/pika/connection.py", line 710, in channel
            self._channels[channel_number].open()
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/pika/channel.py", line 635, in open
            self._rpc(spec.Channel.Open(), self._on_openok, [spec.Channel.OpenOk])
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/pika/channel.py", line 1139, in _rpc
            self._send_method(method_frame)
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/pika/channel.py", line 1150, in _send_method
            self.connection._send_method(self.channel_number, method_frame, content)
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/pika/connection.py", line 1569, in _send_method
            self._send_frame(frame.Method(channel_number, method_frame))
          File "/home/vagrant/.virtualenvs/pika/lib/python2.7/site-packages/pika/connection.py", line 1548, in _send_frame
            raise exceptions.ConnectionClosed
        pika.exceptions.ConnectionClosed: 
2016-11-01T19:54:30+0000 [twisted.application.internet.ClientService#info] Scheduling retry 1 to connect <twisted.internet.endpoints.TCP4ClientEndpoint object at 0x7f07a05603d0> in 2.15385893877 seconds.
2016-11-01T19:54:30+0000 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7f07a0508b48>
2016-11-01T19:54:32+0000 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7f07a0508b48>
2016-11-01T19:54:32+0000 [twisted.application.internet.ClientService#info] Scheduling retry 2 to connect <twisted.internet.endpoints.TCP4ClientEndpoint object at 0x7f07a05603d0> in 3.20834667501 seconds.
2016-11-01T19:54:32+0000 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7f07a0508b48>
2016-11-01T19:54:35+0000 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7f07a0508b48>
2016-11-01T19:54:35+0000 [twisted.application.internet.ClientService#info] Scheduling retry 3 to connect <twisted.internet.endpoints.TCP4ClientEndpoint object at 0x7f07a05603d0> in 3.5323085557 seconds.
2016-11-01T19:54:35+0000 [twisted.internet.protocol.Factory#info] Stopping factory <twisted.internet.protocol.Factory instance at 0x7f07a0508b48>
2016-11-01T19:54:39+0000 [twisted.internet.protocol.Factory#info] Starting factory <twisted.internet.protocol.Factory instance at 0x7f07a0508b48>
```

Basically, Pika is upset for trying to get a channel from a connection that's closed. Am I mis-reading how ``whenConnected`` should work?

**Comment by glyph** [comment]
I don't know how Pika is attempting to keep track of connection state, but with simpler tests I can't reproduce this.  `whenConnected` seems to do roughly the right thing and reconnect.  Can you boil this down to a more minimal (hopefully no pika) reproducer?

**Comment by vitaly-krugl** [comment]
`pika.connection.Connection` calls `on_closed_callback` that is passed to the constructor or registered via `Connection.add_on_close_callback()`. The `raise exceptions.ConnectionClosed` in `self._send_frame` indicates that the app didn't register or ignored `on_closed_callback` and continued to use the connection after that either directly or indirectly.

**Comment by vitaly-krugl** [comment]
Issue has been stale for a while and based on the back-and-forth appears to be user error. Closing. Please reopen if you have new details and a simpler reproducible example to share.

**Comment by vitaly-krugl** [comment]
@glyph, @jeremycline - if you're interested in preserving Twisted support in Pika, please see issue #1031.
